### PR TITLE
[AMD] Support AMD MXFP4 Qwen3.5-397B-A17B model

### PR DESCRIPTION
**Pull request by hubertlu-tw**
## Motivation

Enable and validate Qwen3.5 MXFP4 model support on AMD GPUs
This PR aims to preserve acceptable accuracy while improving serving performance versus the FP8 baseline.

## Modifications

- Added SGLang changes to support [amd/Qwen3.5-397B-A17B-MXFP4](https://huggingface.co/amd/Qwen3.5-397B-A17B-MXFP4) inference on AMD GPUs.

<details>
<summary>Server command: FP8 baseline</summary>

```bash
SGLANG_USE_AITER=1 python3 -m sglang.launch_server \
  --model-path Qwen/Qwen3.5-397B-A17B-FP8 \
  --tp 4 \
  --attention-backend aiter \
  --trust-remote-code \
  --model-loader-extra-config '{"enable_multithread_load": true}' \
  --watchdog-timeout 1200 \
  --mem-fraction-static 0.9 \
  --host 0.0.0.0 --port 9000
```

</details>

<details>
<summary>Server command: MXFP4 model</summary>

```bash
SGLANG_USE_AITER=1 python3 -m sglang.launch_server \
  --model-path amd/Qwen3.5-397B-A17B-MXFP4 \
  --tp 4 \
  --attention-backend aiter \
  --trust-remote-code \
  --model-loader-extra-config '{"enable_multithread_load": true}' \
  --watchdog-timeout 1200 \
  --mem-fraction-static 0.9 \
  --host 0.0.0.0 --port 9000
```

</details>

## Accuracy Tests

Accuracy command:

```bash
python3 benchmark/gsm8k/bench_sglang.py \
  --num-questions 1319 --parallel 1319 --num-shots 5 --port 9000
```

### GSM8K results (FP8 vs MXFP4)

| Model | Run | Accuracy | Invalid | Latency (s) | Output Throughput (tok/s) |
|---|---:|---:|---:|---:|---:|
| FP8 | 1 | 0.951 | 0.009 | 78.487 | 2768.601 |
| FP8 | 2 | 0.948 | 0.008 | 79.401 | 2756.652 |
| MXFP4 | 1 | 0.931 | 0.004 | 68.742 | 3228.576 |
| MXFP4 | 2 | 0.932 | 0.004 | 71.928 | 3125.804 |

### Averaged GSM8K summary

| Model | Avg Accuracy | Avg Invalid | Avg Latency (s) | Avg Throughput (tok/s) |
|---|---:|---:|---:|---:|
| FP8 | 0.9495 | 0.0085 | 78.9440 | 2762.627 |
| MXFP4 | 0.9315 | 0.0040 | 70.3350 | 3177.190 |

## Benchmarking and Profiling

Serving benchmark command:

```bash
python3 -m sglang.bench_serving \
  --host localhost --port 9000 \
  --model /data2/Qwen/Qwen3.5-397B-A17B-FP8/ \
  --dataset-name random \
  --random-input 8192 --random-output 1024 \
  --random-range-ratio 1.0 \
  --max-concurrency 1 --num-prompt 8
```

### Latency and Throughput

| Metric | TP4 (FP8) | TP4 (MXFP4) | Improvement (TP4 vs TP4 MXFP4) |
|---|---:|---:|---:|
| Median E2E Latency (ms) | 11,233.30 | 9,508.52 | 15.35% |
| Total Throughput (tok/s) | 820.2 (205.05/GPU) | 968.6 (242.15/GPU) | 18.09% |

### TTFT and ITL

| Metric | TP4 (FP8) | TP4 (MXFP4) | Improvement (TP4 vs TP4 MXFP4) |
|---|---:|---:|---:|
| Median TTFT (ms) | 289.27 | 230.91 | 20.17% |
| Median ITL (ms) | 10.70 | 9.05 | 15.42% |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

CC: @HaiShaw @kkHuang-amd @yichiche

### CODE REVIEW

File: python/sglang/srt/models/qwen3_5.py [hunk 1]
```diff
@@ -891,6 +902,16 @@ def __init__(
         self.hidden_size = config.hidden_size
         self.pp_group = get_pp_group()
 
+        # Quark checkpoints provide exclusion lists in
+        # unfused module names. Populate packed mapping so skip checks can
+        # unfold fused SGLang names into Quark-compatible names.
+        if (
+            _is_hip
```
Comment (BowenBao) [inline python/sglang/srt/models/qwen3_5.py]: is _is_hip required or can this be relaxed for all quant_config?

File: python/sglang/srt/models/qwen3_5.py [hunk 2]
```diff
@@ -1285,6 +1306,11 @@ def load_fused_expert_weights(
 
 
 class Qwen3_5ForConditionalGeneration(Qwen3VLForConditionalGeneration):
+    packed_modules_mapping = (
+        Qwen3_5ForCausalLM.packed_modules_mapping if _is_hip else {}
```
Comment (BowenBao) [inline python/sglang/srt/models/qwen3_5.py]: same here, could it be relaxed without `_is_hip` check?

File: python/sglang/srt/models/qwen3_5.py (hunk 1 above)
Comment (hubertlu-tw) [reply python/sglang/srt/models/qwen3_5.py]: This is not required just for the sake of keeping other code paths untouched for non AMD GPUs. @HaiShaw any thoughts on this?

**Review by HaiShaw — COMMENTED** [review]
Please isolate quark incurred code changes, either move to a separate/utility file identifiable with `_quark` in name, or within `_is_hip`, though 1st approach looks more clean.

File: python/sglang/srt/models/qwen3_5.py (hunk 1 above)
Comment (hubertlu-tw) [reply python/sglang/srt/models/qwen3_5.py]: @BowenBao I have removed is_hip check and make it hardware-agnostic if the quark's quant is used.
Could you please review the code changes again? Thanks!

File: python/sglang/srt/models/qwen3_5.py [hunk 3]
```diff
@@ -879,6 +882,18 @@ def forward(
 class Qwen3_5ForCausalLM(nn.Module):
     """Qwen3.5 Model with support for dense variant."""
 
+    # Keep default mapping behavior for non-quark quantization.
+    packed_modules_mapping = {
+        "qkv_proj": ["q_proj", "k_proj", "v_proj"],
+        "gate_up_proj": ["gate_proj", "up_proj"],
+    }
+    # Quark-only mapping used to unfold fused names for exclude matching.
+    quark_packed_modules_mapping = {
```
Comment (BowenBao) [inline python/sglang/srt/models/qwen3_5.py]: imo this is not quark specific, if we look at official fp8 model from qwen https://huggingface.co/Qwen/Qwen3.5-397B-A17B-FP8/blob/main/config.json#L148 'in_proj_ba' is also split. 'in_proj_qkvz' is split as well but they were not excluded by the fp8 config.

Image

does the fp8 model work with upstream?

Comment (BowenBao) [reply python/sglang/srt/models/qwen3_5.py]: perhaps in general all quantization would require the module split info. I'm fine with keeping quark specific if the fp8 already works with upstream.

**Review by HaiShaw — APPROVED** [review]
@hubertlu-tw @yichiche @BowenBao 
Please open another issue/PR to track accuracy.
